paddle/fluid/operators/fused/mkldnn/multi_gru_mkldnn_op.cc

```diff
@@ -0,0 +1,694 @@
+
+  // Reorder output values to PP format [N, T, C] -> [WORDS, C]
+  void reorderNTCtoPP(void* input_data, void* output_data, int layer) {
+    auto* input_data_iter = reinterpret_cast<T_out*>(input_data);
+    auto* output_data_iter = reinterpret_cast<T_out*>(output_data);
+    auto oc = OCs[layer] * 2;
+    for (int n = 0; n < N_; ++n) {
+      const auto num_elements = (x_lod_[n + 1] - x_lod_[n]) * oc;
+      memcpy(output_data_iter, input_data_iter + n * Ti_ * oc,
+             sizeof(T_out) * num_elements);
+      output_data_iter += num_elements;
+    }
+  }
+
+  // Reorder output values to PP format [T, N, C] -> [WORDS, C]
+  void reorderTNCtoPP(void* input_data, void* output_data, int layer) {
+    auto* input_data_iter = reinterpret_cast<T_out*>(input_data);
+    auto* output_data_iter = reinterpret_cast<T_out*>(output_data);
+    for (int n = 0; n < N_; ++n) {
+      const auto num_elements = x_lod_[n + 1] - x_lod_[n];
+      for (size_t t = 0; t < num_elements; ++t) {
+        memcpy(output_data_iter,
+               input_data_iter + t * N_ * OCs[layer] + n * OCs[layer],
+               sizeof(T_out) * OCs[layer]);
+        output_data_iter += OCs[layer];
+      }
+    }
+  }
+
+ private:
+  // RNN dimensions
+  // N - Batch Size
+  // Ti - Max sentence length
+  // ICs - Input Channels
+  // OCs - Output Channels
+  int64_t N_, Ti_;
+  std::vector<int64_t> ICs, OCs;
+
+  const platform::MKLDNNDeviceContext& dev_ctx_;
+  const dnnl::engine engine_;
+  const platform::Place place_;
+  const bool origin_mode_;
+  const int layers_;
+
+  std::map<std::pair<int, Direction>,
+           std::shared_ptr<dnnl::gru_forward::primitive_desc>>
+      gru_pds_;
+  std::vector<std::shared_ptr<dnnl::concat::primitive_desc>> concat_pds_;
+
+  std::string key_;
+  // Memory size of weights, bias and h0 does not depend
+  // on Ti size, thus we need another key to cache them
+  std::string memory_key_;
+
+  const LoDTensor* x_;
+  const std::vector<const Tensor*> weights_x_;
+  const std::vector<const Tensor*> weights_h_;
+  const std::vector<const Tensor*> biases_;
+  LoDTensor* hidden_;
+  std::vector<dnnl::primitive_attr> attrs_;
+  const paddle::framework::Vector<size_t>& x_lod_;
+};
+
+template <typename T>
+class MultiGRUMKLDNNKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    const bool force_fp32_output =
+        ctx.HasAttr("force_fp32_output") && ctx.Attr<bool>("force_fp32_output");
+
+    if (force_fp32_output) {
+      RunKernel<float>(ctx);
+    } else {
+      RunKernel<T>(ctx);
+    }
+  }
+
+  template <typename Tout = T>
+  void RunKernel(const framework::ExecutionContext& ctx) const {
+    auto& dev_ctx =
+        ctx.template device_context<platform::MKLDNNDeviceContext>();
+    MultiGRUHandler<T, Tout> handler(ctx, dev_ctx);
+
+    int layers = handler.getLayers();
+    auto input_mem = handler.AcquireInputMemoryWithReorder();
```

Comment:
So this code is basically executing GRU units independently and then concatenates the results. Why oneDNN RNN cell "bidirectional_concat" cannot be used in this situation ?

Reply:
The oneDNN `bidirectional_concat` version of GRU could be used only for bs=1. For bs>1 transformation of input data from Paddle format to oneDNN format requires adding padding and the `bidirectional_concat` GRU cannot handle that properly.

Reply:
ok